client/asset/dcr/externaltx.go

```diff
@@ -0,0 +1,444 @@
+	// process got here first, a search for the tx output spender might
+	// be underway already. Wait until any previous search completes and
+	// releases this lock.
+	output.spender.scanMtx.Lock()
+	defer output.spender.scanMtx.Unlock()
+
+	// Check if the spender of this tx output was found in a previous search
+	// attempt.
+	spenderBlockHash, err := output.spender.relevantBlockHash(dcr.getBlockHash)
+	if err != nil {
+		return false, err
+	}
+	if spenderBlockHash != nil {
+		return true, nil
+	}
+
+	// This tx output is not known to be spent as of last search (if any).
+	// Scan blocks from the lastScannedBlock (if there was a previous scan)
+	// or from the tx block to attempt finding the spender of this output.
+	// Use mainChainAncestor to ensure that scanning starts from a mainchain
+	// block in the event that either tx block or lastScannedBlock have been
+	// re-orged out of the mainchain.
+	var startBlock *chainjson.GetBlockVerboseResult
+	if output.spender.lastScannedBlock == nil {
+		_, startBlock, err = dcr.mainChainAncestor(txBlockHash)
+	} else {
+		_, startBlock, err = dcr.mainChainAncestor(output.spender.lastScannedBlock)
+	}
+	if err != nil {
+		return false, err
+	}
+
+	// Search for this output's spender in the blocks between startBlock and bestBlock.
+	bestBlock := dcr.cachedBestBlock()
+	dcr.log.Debugf("Searching for the tx that spends output %s in blocks %d (%s) to %d (%s).",
+		output, startBlock.Height, startBlock.Hash, bestBlock.height, bestBlock.hash)
+	for blockHeight := startBlock.Height; blockHeight <= bestBlock.height; blockHeight++ {
+		blockHash, err := dcr.node.GetBlockHash(dcr.ctx, blockHeight)
+		if err != nil {
+			return false, translateRPCCancelErr(err)
+		}
+		blockFilter, err := dcr.getBlockFilterV2(blockHash)
+		if err != nil {
+			return false, err
+		}
+		if !blockFilter.Match(output.PkScript) {
+			output.spender.lastScannedBlock = blockHash
+			continue // check next block's filters (blockHeight++)
+		}
+		dcr.log.Debugf("Block %d (%s) likely contains a tx that spends %s. Confirming.",
+			blockHeight, blockHash, output)
+		blk, err := dcr.getBlock(blockHash, true)
+		if err != nil {
+			return false, err
+		}
+		for i := range blk.RawTx {
+			blkTx := &blk.RawTx[i]
+			if txSpendsOutput(blkTx, output) {
+				dcr.log.Debugf("Found spender for output %s in block %d (%s), spender tx hash %s.",
+					output, blockHeight, blockHash, blkTx.Txid)
+				output.spender.relevantBlock = &block{hash: blockHash, height: blockHeight}
+				return true, nil
+			}
+		}
+		for i := range blk.RawSTx {
+			blkTx := &blk.RawSTx[i]
+			if txSpendsOutput(blkTx, output) {
+				dcr.log.Debugf("Found spender for output %s in block %d (%s), spender tx hash %s.",
+					output, blockHeight, blockHash, blkTx.Txid)
+				output.spender.relevantBlock = &block{hash: blockHash, height: blockHeight}
+				return true, nil
+			}
+		}
+		output.spender.lastScannedBlock = blockHash
```

Comment:
Getting here should be fairly rare according to the GCS parameters, so I think this is worth a log.  If we see this hit and logged often, something is probably wrong.